daemon/ipcache.go

```diff
@@ -0,0 +1,147 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/api/v1/models"
+	. "github.com/cilium/cilium/api/v1/server/restapi/policy"
+	"github.com/cilium/cilium/pkg/api"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
+
+	"github.com/go-openapi/runtime/middleware"
+)
+
+// OnIPIdentityCacheChange listens to ipcache updates and forwards them to the monitor
+func (d *Daemon) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr net.IPNet,
+	oldHostIP, newHostIP net.IP, oldID *identity.NumericIdentity, newID identity.NumericIdentity,
+	encryptKey uint8, k8sMeta *ipcache.K8sMetadata) {
+	var (
+		k8sNamespace, k8sPodName string
+		newIdentity, oldIdentity uint32
+		oldIdentityPtr           *uint32
+	)
+
+	if k8sMeta != nil {
+		k8sNamespace = k8sMeta.Namespace
+		k8sPodName = k8sMeta.PodName
+	}
+
+	newIdentity = newID.Uint32()
+	if oldID != nil {
+		oldIdentity = (*oldID).Uint32()
+		oldIdentityPtr = &oldIdentity
+	}
+
+	repr, err := monitorAPI.IPCacheNotificationRepr(cidr.String(), newIdentity, oldIdentityPtr,
+		newHostIP, oldHostIP, encryptKey, k8sNamespace, k8sPodName)
+	if err == nil {
+		switch modType {
+		case ipcache.Upsert:
+			d.SendNotification(monitorAPI.AgentNotifyIPCacheUpserted, repr)
+		case ipcache.Delete:
+			d.SendNotification(monitorAPI.AgentNotifyIPCacheDeleted, repr)
+		}
+	}
+}
+
+// OnIPIdentityCacheGC is required to implement IPIdentityMappingListener.
+func (d *Daemon) OnIPIdentityCacheGC() {
+	// Nothing to do, this event is currently not useful to monitor clients
+}
+
+type getIPCache struct {
+	daemon *Daemon
+}
+
+// NewGetIPCacheHandler for the global IP cache
+func NewGetIPCacheHandler(d *Daemon) GetIPCacheHandler {
+	return &getIPCache{daemon: d}
+}
+
+func (h *getIPCache) Handle(params GetIPCacheParams) middleware.Responder {
+	listener := &ipCacheDumpListener{}
+	if params.Cidr != nil {
+		_, cidrFilter, err := net.ParseCIDR(*params.Cidr)
+		if err != nil {
+			return api.Error(GetIPCacheBadRequestCode, err)
+		}
+		listener.cidrFilter = cidrFilter
+	}
+	ipcache.IPIdentityCache.RLock()
+	ipcache.IPIdentityCache.DumpToListenerLocked(listener)
+	ipcache.IPIdentityCache.RUnlock()
+	if len(listener.entries) == 0 {
+		return NewGetIPCacheNotFound()
+	}
+
+	return NewGetIPCacheOK().WithPayload(&models.IPCache{
+		Cache: listener.entries,
```

Comment:
(Are there similar examples of large dumps we may do in the API already?)

Reply:
That's a very fair point. There is no maximum length, but there are similar examples: I modeled the API to roughly mirror `/fqdn/cache`, which also does not have a limit. But I'm not sure how the IP cache compares to the FQDN cache in terms of size and update rates. While it depends on the workload, I'm guessing the IP cache will have at least as many entries and updates as the FQDN cache?

In terms of impact, dumping the cache via API acquires a read-lock, which naturally will block any IP cache updates. Not sure what we can do here to avoid . My use case for the API is to obtain a initial snapshot of the IP cache. 

Thinking about it, splitting the API into the typical `list keys` (which should be cheaper) and `fetch details for key` endpoints will allow updates to be interleaved. This will of course (by design) not give me complete snapshot of the cache, but I should be fine since I can listen to monitor events and deal with invalidated entries accordingly.

Reply:
After looking at other API endpoints, I'm less sure about splitting the API.

All other collection paths (`/endpoint`, `/identity`, `/service`, the aforementioned `/fqdn/cache`) dump fully populated lists upon request. They all arguably all contain less entries than the IP cache, but the response for some (e.g. `/endpoint`) will likely be larger, as they contain more fields.

Reply:
FQDN cache is a good question; I forget whether these are generated based upon the number of FQDN rules in the policy, or based upon the number of unique FQDNs seen by the DNS proxy. I suspect it's the former, which would limit it considerably.

Looking at endpoint, I don't think we're expecting more than perhaps 100 endpoints on a given node (which even that is quite high compared to typical deployments today). Looking at the ipcache, the default maximum is 512,000. While the endpoint is a more complicated structure, the order of magnitude is very different.

Identities is probably closer to the scale of ipcache, while I don't think that there is a BPF map that tells us the limit, we've investigated in the order of tens of thousands.

One thing you might consider doing to test this out and just generally make sure the API behaves correctly is to inject high numbers of IP-ID entries into the kvstore. I did something like that here: https://github.com/cilium/cilium/issues/8063#issuecomment-493236355

The read-lock question is mainly interesting if it's going to happen particularly often. Do you have a sense for your use case how frequently you want to hit that API? There are mechanisms to make that behaviour better if necessary, but it's quite a bit more work.

Reply:
Given that we only hold the lock to dump into memory, I think it is fine.

Reply:
> One thing you might consider doing to test this out and just generally make sure the API behaves correctly is to inject high numbers of IP-ID entries into the kvstore. I did something like that here: #8063 (comment)

Oh, that's a great idea. I'll give it a shot.

> Do you have a sense for your use case how frequently you want to hit that API? There are mechanisms to make that behaviour better if necessary, but it's quite a bit more work.

My goal is to maintain a history of the ipcache in downstream consumers of the cilium monitor socket. I think we can reduce the API load quite a if we rely on monitor socket notifications. If I fetch the ipcache snapshot after connecting to the monitor socket I should be able maintain my copy of the ipcache via notifications only and do not need to consult the API at all after that (until the monitor socket gets disconnected for whatever reason of course). 

(This assumes I can detect update notifications older than my initial snapshot as such, which I think I can since the upserts contain the old values they replace). If not, we would have to add a node-local revision field to the ipcache)

Reply:
@joestringer I ran a rough benchmark yesterday to get an idea how expensive the RLock is, the numbers are from a VM running on my dev laptop, so take them with a grain of salt as there was likely some background noise.

I ran the experiment with 25, 50 and 75k entries in the ipcache. I was a bit surprised to see that it is indeed not negligible. With 75k ipcache entries, the lock is held on average 92 milliseconds. 

Percentiles (all values in milliseconds):

| IPCache Entries:      |         25k |      50k |      75k |
|:------|------------:|---------:|---------:|
| mean  |    25.2579  |  56.7184 |  92.4482 |
| std   |     7.55403 |  10.8656 |  22.2895 |
| 50%   |    21.8291  |  51.7066 |  86.0281 |
| 75%   |    28.3468  |  64.2931 | 103.095  |
| 99.9% |    66.3948  | 143.382  | 227.67   |
| 100%  |   422.616   | 175.278  | 641.549  |

Boxplot (values larger than 99.8%-percentile removed for readability):
![image](https://user-images.githubusercontent.com/50564/65955968-97adb500-e449-11e9-96df-af2641913d1b.png)

In conclusion, it can block ipcache updates for hunders of millisconds when using the API. In really bad cases likely more a second if we have >100k entries. As discussed, we will have to be careful with calling the new `/ip/` API endpoint to not add unnecessarily block updates.